packages/worker/src/workflow/vm.ts

```diff
@@ -0,0 +1,147 @@
+  public static async create(code: string, isolateExecutionTimeoutMs: number): Promise<VMWorkflowCreator> {
+    const script = new vm.Script(code, { filename: 'workflow-isolate' });
+    return new this(script, isolateExecutionTimeoutMs);
+  }
+
+  public async destroy(): Promise<void> {
+    delete this.script;
+  }
+}
+
+type WorkflowModule = typeof internals;
+
+export class VMWorkflow implements Workflow {
+  constructor(
+    public readonly info: WorkflowInfo,
+    protected context: vm.Context | undefined,
+    readonly workflowModule: WorkflowModule,
+    public readonly isolateExecutionTimeoutMs: number
+  ) {}
+
+  async getAndResetSinkCalls(): Promise<SinkCall[]> {
+    return this.workflowModule.getAndResetSinkCalls();
+  }
+
+  /**
+   * Inject a function into the isolate context global scope
+   *
+   * @param path name of global variable to inject the function as (e.g. `console.log`)
+   * @param fn function to inject into the isolate
+   */
+  public async injectGlobal(key: string, val: unknown): Promise<void> {
+    if (this.context === undefined) {
+      throw new IllegalStateError('Workflow isolate context uninitialized');
+    }
+    this.context[key] = val;
+  }
+
+  public async activate(activation: coresdk.workflow_activation.IWFActivation): Promise<Uint8Array> {
+    if (this.context === undefined) {
+      throw new IllegalStateError('Workflow isolate context uninitialized');
+    }
+    this.info.isReplaying = activation.isReplaying ?? false;
+    if (!activation.jobs) {
+      throw new Error('Expected workflow activation jobs to be defined');
+    }
+
+    // Job processing order
+    // 1. patch notifications
+    // 2. signals
+    // 3. anything left except for queries
+    // 4. queries
+    const [patches, nonPatches] = partition(activation.jobs, ({ notifyHasPatch }) => notifyHasPatch != null);
+    const [signals, nonSignals] = partition(nonPatches, ({ signalWorkflow }) => signalWorkflow != null);
+    const [queries, rest] = partition(nonSignals, ({ queryWorkflow }) => queryWorkflow != null);
+    let batchIndex = 0;
+
+    // Loop and invoke each batch and wait for microtasks to complete.
+    // This is done outside of the isolate because when we used isolated-vm we couldn't wait for microtasks from inside the isolate, not relevant anymore.
+    for (const jobs of [patches, signals, rest, queries]) {
+      if (jobs.length === 0) {
+        continue;
+      }
+      const arr = coresdk.workflow_activation.WFActivation.encodeDelimited({ ...activation, jobs }).finish();
+      const { numBlockedConditions } = await this.workflowModule.activate(arr, batchIndex++);
+      if (numBlockedConditions > 0) {
+        await this.tryUnblockConditions();
+      }
+      // Wait for microtasks to be processed
+      await new Promise((resolve) => process.nextTick(resolve));
+    }
+    return this.workflowModule.concludeActivation();
+  }
+
+  protected async tryUnblockConditions(): Promise<void> {
+    for (;;) {
```

Comment:
`dispose` will not be called if the workflow is still activating in this version of our runtime.

I tried to think of a way where workflow code creates new conditions which are unblocked in an endless loop but I can't find a way to do that because the condition function is eagerly evaluated in the WF isolate.